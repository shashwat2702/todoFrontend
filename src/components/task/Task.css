.individualTask {
    width:100%;
}
.existingTodoTask {
    background-color: #222222;
    color: gray;
    border: 0;
    font-size: 1.5rem;
    outline: none;
    border-bottom: 1px solid white;
    padding: 0.5rem 0;
    width: 100%;
}
.editableTodoTask {
    background-color: #222222;
    color: white;
    border: 0;
    font-size: 1.5rem;
    outline: none;
    border-bottom: 1px solid white;
    padding: 0.5rem 0;
    width: 100%;
}
.editTaskButton {
    background-color: dodgerblue;
    width: 12rem;
    height: 2rem;
    border: 0;
    border-radius: 0.2rem;
    margin: 1rem 1rem 0 0;
    font-size: 1.5rem;
    color: white;
}
.removeTaskButton {
    background-color: dodgerblue;
    width: 12rem;
    height: 2rem;
    border: 0;
    border-radius: 0.2rem;
    margin-top: 1rem;
    font-size: 1.5rem;
    color: white;
}