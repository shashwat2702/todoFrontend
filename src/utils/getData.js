const axios = require('axios');

const getData = url => axios.get(url);
const postData = (url, data) => axios({
  method: 'post',
  url,
  data,
});
export { getData, postData };
