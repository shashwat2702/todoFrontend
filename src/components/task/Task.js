/* eslint-disable no-shadow */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { updateData } from '../../utils/getData';
import './Task.css';

export default class Task extends Component {
  constructor(props) {
    super(props);
    this.state = {
      readonly: true,
      task: props.task,
      taskId: props.id,
    };
  }

  makeEditable = () => {
    this.setState({ readonly: false });
  }

  handleInputChange = (event) => {
    const { target } = event;
    const { value } = target;
    const { name } = target;
    this.setState({
      [name]: value,
    });
  }

  updateTask = (event) => {
    event.preventDefault();
    const { task, taskId } = this.state;
    updateData('http://localhost:8080/updateTask', taskId, task)
      .then(({ data }) => {
        if (data.updated === true) {
          this.setState({ task: data.task });
        } else {
          const { task } = this.props;
          this.setState({ task });
        }
      },
      () => {
        const { task } = this.props;
        this.setState({ task });
      });
    this.setState({ readonly: true });
  };

  render() {
    const { readonly, task, taskId } = this.state;
    const { onDelete } = this.props;
    return (
      <div className="individualTask">
        <form onSubmit={(e) => { e.preventDefault(); }}>
          <input
            value={task}
            readOnly={readonly}
            className={readonly ? 'existingTodoTask' : 'editableTodoTask'}
            name="task"
            id={taskId}
            onChange={this.handleInputChange}
          />
          <br />
          {!readonly
        && (
        <button className="editTaskButton" type="submit" onClick={this.updateTask}>
            Update Task
        </button>
        )
        }
          {readonly
        && (
        <button className="editTaskButton" type="submit" onClick={this.makeEditable}>
            Edit Task
        </button>
        )
        }
          <button className="removeTaskButton" id={taskId} type="submit" onClick={onDelete}>
            Remove Task
          </button>
        </form>
      </div>
    );
  }
}

Task.propTypes = {
  task: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  onDelete: PropTypes.func.isRequired,
};
