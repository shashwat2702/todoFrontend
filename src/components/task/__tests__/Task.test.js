import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Task from '../Task';
import * as apiCalls from '../../../utils/getData';

describe('Task snapshot', () => {
  it('should render correctly', () => {
    const TaskSnap = renderer.create(<Task
      task="Create Todo"
      id={1}
      onDelete={jest.fn()}
    />);
    const TaskJson = TaskSnap.toJSON();
    expect(TaskJson).toMatchSnapshot();
  });
});

describe('makeEditable', () => {
  it('should set readonly attribute inside state to false', () => {
    jest.spyOn(apiCalls, 'getData');
    const wrapper = shallow(<Task
      task="Create Todo"
      id={1}
      onDelete={jest.fn()}
    />);
    const instance = wrapper.instance();
    expect(instance.state.readonly).toEqual(true);
    instance.makeEditable();
    expect(instance.state.readonly).toEqual(false);
  });
});

describe('handleInputChange', () => {
  it('should set value in task attribute of state', () => {
    const wrapper = shallow(<Task
      task="Create Todo"
      id={1}
      onDelete={jest.fn()}
    />);
    const instance = wrapper.instance();
    const event = {
      target: {
        name: 'task',
        value: 'Create',
      },
    };
    instance.handleInputChange(event);
    expect(instance.state.task).toEqual(event.target.value);
  });
});

describe('updateTask', () => {
  it('should call updateData with parameter taskId and new task', () => {
    jest.spyOn(apiCalls, 'updateData');
    const wrapper = shallow(<Task
      task="Create Todo"
      id={1}
      onDelete={jest.fn()}
    />);
    const event = {
      preventDefault: jest.fn(),
    };
    const newState = { task: 'Updated Task', taskId: 1 };
    const instance = wrapper.instance();
    instance.setState(newState);
    instance.updateTask(event);
    expect(apiCalls.updateData).toHaveBeenCalledWith('http://localhost:8080/updateTask', newState.taskId, newState.task);
  });
});
