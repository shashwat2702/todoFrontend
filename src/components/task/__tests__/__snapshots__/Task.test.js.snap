// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task snapshot should render correctly 1`] = `
<div
  className="individualTask"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="existingTodoTask"
      id={1}
      name="task"
      onChange={[Function]}
      readOnly={true}
      value="Create Todo"
    />
    <br />
    <button
      className="editTaskButton"
      onClick={[Function]}
      type="submit"
    >
      Edit Task
    </button>
    <button
      className="removeTaskButton"
      id={1}
      onClick={[MockFunction]}
      type="submit"
    >
      Remove Task
    </button>
  </form>
</div>
`;
